### DISCUSSION TITLE
does Pulsar has data repair mechanism like read repair in Cassandra?

### DISCUSSION THREAD

**Discussion by huangydyn** [apache/pulsar#19734] [Q&A] [answered]
As we know in Cassandra, there is read repair which is used to make data consistent in quorum write.

In Pulsar, let's assume that Write Quorum Size (Qw) is 4, Ack Quorum Size (Qa) is 3.

If client try to write msg01 and ack is 2，then this request has been failed. But `msg01` has been wrote successfully in 2 bookie. How does Pulsar repair/rollback this `msg01`?

**Comment by jak78** [comment, marked as answer]
Hi @huangydyn! Pulsar uses BookKeeper as persistence storage, and BookKeeper has an _AutoRecovery_ mechanism. It is responsible for bringing the cluster back to full replication. You will find the documentation here: https://bookkeeper.apache.org/docs/admin/autorecovery/ 
You can also watch this video to know more: https://www.youtube.com/watch?v=w14OoOUkyvo
